scripts/repair_batlas_references.py

```diff
@@ -0,0 +1,189 @@
+        raise
+
+
+def chunked_iterable(iterable, size):
+    """Loop over an iterable and return chunks of @size"""
+    it = iter(iterable)
+    while True:
+        chunk = tuple(itertools.islice(it, size))
+        if not chunk:
+            break
+        yield chunk
+
+
+def translate_to_citation(json_record):
+    json_key_to_citation_key = {
+        "accessURI": "access_uri",
+        "alternateURI": "alternate_uri",
+        "bibliographicURI": "bibliographic_uri",
+        "citationDetail": "citation_detail",
+        "formattedCitation": "formatted_citation",
+        "otherIdentifier": "identifier",
+        "shortTitle": "short_title",
+        "type": "type",
+    }
+
+    as_citation = {}
+    for key, value in json_record.iteritems():
+        key = key.strip()
+        value = value.strip()
+        new_key = json_key_to_citation_key[key]
+        if isinstance(value, unicode):
+            new_value = value.encode("utf-8")
+        else:
+            new_value = value
+
+        as_citation[new_key] = new_value
+
+    return as_citation
+
+
+def diff_citations(model, candidate):
+    """Return keys where the values differ after whitespace is stripped.
+
+    Assume both dicts have the same keys.
+    """
+    return [key for key in model.keys() if model[key].strip() != candidate[key].strip()]
+
+
+def update_citation(place, index, new_citation_value):
+    citations = place.getReferenceCitations()
+    citations[index] = new_citation_value
+
+    place.setReferenceCitations(citations)
+
+
+def update_places_from_json(container, json_data):
+    results = {
+        "successes": [],
+        "problems": [],
+    }
+    container_url = container.absolute_url()
+
+    for place_id, data in json_data:
+        # Do we have a Place with this ID?
+        place = container.get(place_id)
+        if place is None:
+            results["problems"].append(
+                {
+                    "ID": place_id,
+                    "msg": "Place ID not found in {}".format(container_url),
+                }
+            )
+            continue
+
+        # Is there a citation at the expected index?
+        expected_index = int(data["ref_position"])
+        try:
+            citation = place.getReferenceCitations()[expected_index]
+        except IndexError:
+            results["problems"].append(
+                {
+                    "ID": place_id,
+                    "msg": u"No citation with index {}".format(expected_index),
+                }
+            )
+            continue
+
+        old = translate_to_citation(data["old"])
+        new = translate_to_citation(data["new"])
+
+        # Does the current data match what was predicted?
+        discrepancies = diff_citations(old, citation)
+        if discrepancies:
+            # Already updated?
+            differences = diff_citations(new, citation)
+            if not differences:
+                msg = "Citation already matches new data"
```

Comment:
I probably wouldn't file this under problems, but stash it someplace else and provide a count of these already updated entries.

Reply:
I think we discussed this and I forgot. Much better:
```
UPDATES: 2 records were updated successfully!
REDUNDANT: 24,115 records were skipped because changes were already applied.
PROBLEMS: 43 records were skipped due to problems:
39432: Data mismatch on 'old' values: ['formatted_citation']
60549: Data mismatch on 'old' values: ['bibliographic_uri', 'formatted_citation', 'short_title', 'citation_detail', 'access_uri']
98908: Data mismatch on 'old' values: ['bibliographic_uri', 'formatted_citation', 'short_title', 'citation_detail', 'access_uri']
207549: Data mismatch on 'old' values: ['bibliographic_uri', 'formatted_citation', 'short_title', 'citation_detail', 'access_uri']
226510: Data mismatch on 'old' values: ['formatted_citation']
...
```